### PR TITLE
Delegated IPAM plugin

### PR DESCRIPTION
**Pull request by wedaly**
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

Add support to Cilium CNI for delegated IPAM plugins as described in the CNI spec: https://www.cni.dev/docs/spec/#delegated-plugin-protocol

I tested this in a [kind cluster](https://kind.sigs.k8s.io/) using the [reference CNI IPAM plugin host-local](https://github.com/containernetworking/plugins/tree/main/plugins/ipam/host-local) to assign IP addresses from `node.spec.PodCIDR`, with `enable-endpoint-health-checking=false` and `local-router-ipv4` set to prevent cilium-daemon from assigning itself an IP. Most of the `cilium connectivity test` suite passed in this configuration, except for the L7 policy tests (not sure why).

I also tested this in an AKS cluster configured to use Cilium with the reference "host-local" delegated plugin for IPAM. In this configuration, all of the Cilium connectivity tests passed.

Fixes: #issue-number

### CODE REVIEW

File: daemon/cmd/daemon_main.go
Comment (joestringer) [inline daemon/cmd/daemon_main.go]: I think it's reasonable to check for something like IPAM == delegated + `option.Config.LocalRouterIPv4 == "" || option.Config.EnableHealthChecking` and fatal out. Probably the better place is in `c *DaemonConfig) Validate()` to do this sort of validation. We probably should think about the `LocalRouterIPv6` option as well, ie how do we make sure that this works in an expected way if either IPv4 or IPv6 is disabled.

It probably wouldn't hurt also for you to poke around in the other usages of these flags just to double-check that there aren't any other expectations for other flags being set when these flags are provided. Mostlikely it's just fine to assume that with delegated IPAM mode, the user must specify these two flags in this particular way though.

File: pkg/ipam/ipam.go
Comment (joestringer) [inline pkg/ipam/ipam.go]: An implementation that returns errors for any attempt to allocate IP addresses would be good IMO. This way at least it isn't a hard-crash.

File: pkg/ipam/option/option.go
Comment (joestringer) [inline pkg/ipam/option/option.go:35]: Worth noting that to use this mode, two pieces of configuration must be provided:

* For `cilium-agent`, `--ipam=delegated-plugin`
* For the CNI configuration, place the delegatee plugin name into the configuration

Given that Cilium's DaemonSet typically installs the CNI configuration file, it may make sense to update the Helm templates to allow the user to specify the IPAM delegation + other plugin in the helm options.

File: plugins/cilium-cni/cilium-cni.go
Comment (joestringer) [inline plugins/cilium-cni/cilium-cni.go]: Probably we should come up with a context early on in this whole function so that we can pass it down to each of these calls consistently and have a deadline that fits within the standard timeouts that e.g. Kubelet might expect. IIRC it's something like... kubelet has a 2 minute timeout so Cilium will try to accomplish everything within ~90s so that Cilium properly catches & implements any failures before returning.

File: plugins/cilium-cni/cilium-cni.go
Comment (joestringer) [inline plugins/cilium-cni/cilium-cni.go]: Seems reasonable enough, this should allow the IPAM specific logic here to be relatively contained.

File: plugins/cilium-cni/cilium-cni.go
Comment (joestringer) [inline plugins/cilium-cni/cilium-cni.go]: Some recent context on this is #18975 (bug where we would unnecessarily churn kernel devices in the case where IPAM is limited). It may actually make sense to move more of the IPAM logic (ie the routes/gateways piece below) above the veth creation logic, so that the code is more self-contained to either invoke this new delegated mode or to figure out IPAM+routes+gateways all together in the standard codepaths of the existing case.

File: plugins/cilium-cni/cilium-cni.go
Comment (joestringer) [inline plugins/cilium-cni/cilium-cni.go]: I'm not sure I quite understand this case. When would `args.Netns` be empty?

File: plugins/cilium-cni/cilium-cni.go
Comment (joestringer) [inline plugins/cilium-cni/cilium-cni.go]: We still need to release any allocated IPs on later failure for the delegated plugin case as well, right?

File: plugins/cilium-cni/cilium-cni.go
Comment (joestringer) [inline plugins/cilium-cni/cilium-cni.go]: Is there a reason we need this in the delete but not in the add case? I would have thought that the CNI binary still has access to the configuration object to be able to detect that delegated IPAM is in use, even from a `CNI DEL` call.

File: daemon/cmd/daemon_main.go
Comment (wedaly) [reply daemon/cmd/daemon_main.go]: Makes sense, thanks!

File: pkg/ipam/option/option.go
Comment (wedaly) [reply pkg/ipam/option/option.go:35]: good call, will take a look at the helm chart as well

File: plugins/cilium-cni/cilium-cni.go
Comment (wedaly) [reply plugins/cilium-cni/cilium-cni.go]: interesting, will take a look.

If I'm understanding the current implementation correctly, the routes are being derived from the host IP address? https://github.com/cilium/cilium/blob/2efbdd68a7a578874d1a9ecb884bfba4f76e0f0b/plugins/cilium-cni/cilium-cni.go#L237-L240 Would it make sense for Cilium CNI to include both that route and any returned by the delegated IPAM plugin?

File: plugins/cilium-cni/cilium-cni.go
Comment (wedaly) [reply plugins/cilium-cni/cilium-cni.go]: I cargo-culted this from the bridge reference plugin. Will check the commit history to see if there's any more context about why they added this.

File: plugins/cilium-cni/cilium-cni.go
Comment (wedaly) [reply plugins/cilium-cni/cilium-cni.go]: there's a `defer` on line 405 above to invoke the delegated plugin's DEL, which should release the IP. I was thinking we'd skip the call to cilium-daemon's `IPAMReleaseIP`, but maybe there's some cleanup there we'd still need?

File: plugins/cilium-cni/cilium-cni.go
Comment (wedaly) [reply plugins/cilium-cni/cilium-cni.go]: yep, the ADD command already had a call to `c.ConfigGet()` to read `conf.DatapathMode`

File: plugins/cilium-cni/cilium-cni.go
Comment (joestringer) [reply plugins/cilium-cni/cilium-cni.go]: These routes would be for inside the pod's namespace, so the main thing is to let it know about the Cilium gateway IP address which is terminated only on the L2 network between the pod and the other side of the veth pair (where Cilium implements an ARP responder). In most cases it should need just a default route with appropriate MTU, pointing towards that IP.

Then depending on the mode (eg `endpointRoutes`), Cilium may separately plumb routes on the host itself.

Comment (wedaly) [reply plugins/cilium-cni/cilium-cni.go]: ah, that makes sense. Thanks!

File: plugins/cilium-cni/cilium-cni.go
Comment (joestringer) [reply plugins/cilium-cni/cilium-cni.go]: Ah right, I just missed this. No, as far as I know this should be good enough.

File: plugins/cilium-cni/cilium-cni.go
Comment (joestringer) [inline plugins/cilium-cni/cilium-cni.go]: Typically we will always use the wrap formatting strings to embed other errors:
```suggestion
		return fmt.Errorf("unable to retrieve configuration from cilium-agent: %w", err)
```

File: plugins/cilium-cni/cilium-cni.go
Comment (joestringer) [inline plugins/cilium-cni/cilium-cni.go]: super minor nit: I think this is actually something like `received empty configuration object from cilium-agent`, otherwise if the agent simply was notable to send the configuration then it would error out in the check above..?

**Comment by wedaly** [comment]
I'm still working on this and plan to update the PR next week.

File: daemon/cmd/daemon_main.go
Comment (wedaly) [reply daemon/cmd/daemon_main.go]: added validation in `DaemonConfig.Validate()`

File: pkg/ipam/ipam.go
Comment (wedaly) [reply pkg/ipam/ipam.go]: added a no-op allocator

File: pkg/ipam/option/option.go
Comment (wedaly) [reply pkg/ipam/option/option.go:35]: I decided not to tackle helm chart / installer changes in this PR, since it's already pretty large. For testing, we've been using the existing helm chart options to install a CNI configuration from a configmap.

File: plugins/cilium-cni/cilium-cni.go
Comment (wedaly) [reply plugins/cilium-cni/cilium-cni.go]: I've left the `context.TODO()`s for now because I didn't want to make changes that could impact non-delegated-plugin codepaths. Interestingly, the bridge reference plugin also uses `context.TODO`, which makes me think it might be relying on the container runtime to kill the CNI process after some timeout.

File: plugins/cilium-cni/cilium-cni.go
Comment (wedaly) [reply plugins/cilium-cni/cilium-cni.go]: updated

File: plugins/cilium-cni/cilium-cni.go
Comment (wedaly) [reply plugins/cilium-cni/cilium-cni.go]: updated

**Comment by wedaly** [comment]
Updated based on the initial round of feedback and testing in an AKS cluster. Planning to validate this with the upstream k/k E2E networking tests later this week, but I think it's far enough along for code review.

File: pkg/ipam/noop_allocator.go
Comment (gandro) [inline pkg/ipam/noop_allocator.go]: tiny nit: The second `string` is used to display the IPAM status in `cilium status`. 

For example, here it's `24/34 allocated`.

```console
$ cilium status
KVStore:                Ok   Disabled
Kubernetes:             Ok   1.21+ (v1.21.9-eks-0d102a7) [linux/amd64]
Cilium:                 Ok   1.10.10 (v1.10.10-024ecfc)
[...]
IPAM:                   IPv4: 24/34 allocated

```

We could return a static string like `"delegated to plugin"` here, instead of not showing anything.

**Review by gandro — APPROVED** [review]
Looks good to me overall! I'm not super familiar with this new CNI feature, but I didn't spot any obvious bugs.

File: pkg/ipam/noop_allocator.go
Comment (wedaly) [reply pkg/ipam/noop_allocator.go]: thanks! will update this

**Comment by wedaly** [comment]
@nathanjsweet @tklauser would you be able to review this? It looks like the PR needs review from a few more code owners. Thanks in advance!

File: pkg/option/config.go
Comment (tklauser) [inline pkg/option/config.go]: For more actionable and user-friendly error messages, I'd suggest to change these to explicitly mention the CLI option names, not the internal names:

```suggestion
		if c.EnableIPv4 && c.LocalRouterIPv4 == "" {
			return fmt.Errorf("--%s must be provided when IPv4 is enabled with --%s=%s", LocalRouterIPv4, IPAM, IPAMDelegatedPlugin)
		}
		if c.EnableIPv6 && c.LocalRouterIPv6 == "" {
			return fmt.Errorf("--%s must be provided when IPv6 is enabled with --%s=%s", LocalRouterIPv6, IPAM, IPAMDelegatedPlugin)
		}
		if c.EnableEndpointHealthChecking {
			return fmt.Errorf("--%s must be disabled with --%s=%s", EnableEndpointHealthChecking, IPAM, IPAMDelegatedPlugin)
		}
```

File: pkg/option/config_test.go
Comment (tklauser) [inline pkg/option/config_test.go]: Checking the actual error for deep equality is not very idiomatic in Go. I'd suggest to just check whether an error is returned or not, i.e. make `expectErr` a `bool` and then change this to something like:

```suggestion
			hasErr := err != nil
			if tt.expectErr && !hasErr {
				t.Errorf("expected error but got none")
			} else if !expectErr && hasErr {
				t.Errorf("expected no error but got %v", err)
			}
```

**Review by tklauser — COMMENTED** [review]
Thanks and sorry for taking so long to review this.

This LGTM apart from the minor comments inline.

File: pkg/option/config_test.go
Comment (wedaly) [reply pkg/option/config_test.go]: thanks, updated!

**Review by tklauser — APPROVED** [review]
Thanks @wedaly

**Comment by wedaly** [comment]
@nathanjsweet would you be able to review this? (If I'm understanding the process correctly, cilium/kubernetes is the last required reviewer). Thanks in advance!

**Comment by wedaly** [comment]
do I need to do anything about the failure in `Cilium L4LB XDP`? From the logs, it looks like `kind` wasn't able to successfully create the cluster ("It seems like the kubelet isn't running or healthy.") Maybe rebasing the branch would help?

File: plugins/cilium-cni/cilium-cni.go
Comment (christarazi) [inline plugins/cilium-cni/cilium-cni.go]: Nit: `%w` for errors

File: plugins/cilium-cni/cilium-cni.go
Comment (christarazi) [inline plugins/cilium-cni/cilium-cni.go:114]: The changes in this commit should probably be squashed with the previous

File: plugins/cilium-cni/cilium-cni.go
Comment (christarazi) [inline plugins/cilium-cni/cilium-cni.go]: `%w`

File: pkg/ipam/noop_allocator.go
Comment (christarazi) [inline pkg/ipam/noop_allocator.go:39]: This commit can also be squashed with the commit that introduced the code

File: pkg/option/config.go
Comment (christarazi) [inline pkg/option/config.go:3409]: And same with this commit re: squashing

**Review by christarazi — CHANGES_REQUESTED** [review]
Changes LGTM overall, noted a few minor things to address. I also noticed that most of the commits don't contain a body describing the "why". Would you mind adding some context into the commits, that way we have the full context in the git history?

**Comment by wedaly** [comment]
> Changes LGTM overall, noted a few minor things to address. I also noticed that most of the commits don't contain a body describing the "why". Would you mind adding some context into the commits, that way we have the full context in the git history?

Thanks! Pushed an updated branch with related commits squashed and "why" descriptions for every commit.

File: plugins/cilium-cni/cilium-cni.go
Comment (christarazi) [reply plugins/cilium-cni/cilium-cni.go]: Still seeing `%s` :slightly_smiling_face:

**Review by christarazi — APPROVED** [review]
LGTM, thanks for adding context to the commits. Please add the copyright header (can be found from other existing files) to the newly added files.

File: plugins/cilium-cni/cilium-cni.go
Comment (christarazi) [reply plugins/cilium-cni/cilium-cni.go]: @wedaly ^

Comment (wedaly) [reply plugins/cilium-cni/cilium-cni.go]: thanks! sorry I missed that, updated!

File: pkg/ipam/option/option.go
Comment (joestringer) [reply pkg/ipam/option/option.go:35]: Makes sense. It would be good to get that documented at some point so others can reproduce and test this setup.

**Comment by joestringer** [comment]
It looks like the smoke test / conformance-test failure is related to https://github.com/cilium/cilium/pull/19987, probably this PR just needs to be rebased against the latest master branch to mitigate. That's OK, we can hold off on that and just watch for the rest of the CI test results.

**Comment by joestringer** [comment]
@wedaly :+1: OK, that's a good sign. Please do rebase, we can check whether the smoke test issue is fixed or has some relation to this PR.

**Pull request merged**
